### ISSUE TITLE
Check if rules already exist in IIS (Windows) web.config file

### ISSUE THREAD

**Issue by markonikolic985** [wp-media/imagify-plugin#509]
When we add rules to IIS config file, we don't check if rules already exist. If it does, it breaks things and users get 500 error.

Example ticket can be found [here](https://secure.helpscout.net/conversation/1260547282/189005?folderId=676708). Once they removed rules manually, it started working.

I haven't seen the full code of their web.config file so maybe we should test it on our end to confirm it will break things by default.

**Comment by markonikolic985** [comment]
There is [another case](https://secure.helpscout.net/conversation/1278290843/193685?folderId=676697). This user shared the code so it might be useful for checking.

**Comment by saranshj4** [comment]
https://secure.helpscout.net/conversation/2641363511/500464/
https://secure.helpscout.net/conversation/2578909533/488158/
https://secure.helpscout.net/conversation/2499684228/472913/
https://secure.helpscout.net/conversation/2546667216/482302/
https://secure.helpscout.net/conversation/2443172507/461813/
https://secure.helpscout.net/conversation/2385590103/447528/
https://secure.helpscout.net/conversation/2156677490/401311/

and few more with tag - img_iis

**Comment by tripflex** [comment]
<img width="596" alt="image" src="https://github.com/user-attachments/assets/8cc4df7d-c259-49a7-b9c7-688dc7404b1d">

You are double loading `staticContent` should only one ever exist

**Comment by tripflex** [comment]
How has this still not been resolved?  This completely takes a site down whenever you activate this

**Comment by tripflex** [comment]
@wordpressfan @MathieuLamiot @Khadreal @jeawhanlee @remyperona

**Comment by markonikolic985** [comment]
Relevant case - https://secure.helpscout.net/conversation/3199226456/615225

**Comment by nicomollet** [comment]
Possible solution: little refactoring of the [IIS classes](https://github.com/search?q=repo%3Awp-media%2Fimagify-plugin%20IIS&type=code) to avoid having `2` `staticContent` blocks, IIS only wants `1` it seems.

**Comment by PlanetThoughts** [comment]
This was first noticed in 2020 and it is still not fixed??  By default, the entire website is crashed, every page. Fortunately I used to work on web implementation and was able to fix it, but what would others do?

**Comment by Honemo** [comment]
> [!NOTE]
> Generated by the AI delivery pipeline (grooming-agent · Claude Opus 4.8).

### Grooming Plan — Issue #509 (Plan v3)

**Approach:** Root-cause fix in `classes/` — emit only a leaf `<mimeMap>` into a single shared `<staticContent>` and dedupe by `@fileExtension` in `AbstractIISDirConfFile::insert_contents()` (before the empty-contents early-return, so it runs on both add/remove). Plus a version-gated, IIS-guarded self-heal migration in `inc/admin/upgrader.php` that collapses Imagify-created duplicate `<staticContent>` siblings on already-broken sites.
**Effort:** L · **Risk:** HIGH · **Complexity:** HIGH

**Changes in v3 (addresses 2nd-round NEEDS_REVISION):**

1. **MUST_HAVE — corrected self-heal re-add gating.** Verified against `classes/Webp/Display.php` + `classes/Avif/Display.php`: **neither reads `optimization_format`.** Both gate purely on `display_nextgen`, and when truthy BOTH the WebP and AVIF subscribers call `->add()` — this is the actual real-world crash trigger (two `<staticContent>` blocks). The v2 `optimization_format` XOR branch was wrong and is dropped. The migration now calls `$webp->add()` AND `$avif->add()` unconditionally when `display_nextgen` is truthy, letting Part 1's dedupe/merge collapse them into one `<staticContent>`.

2. **MUST_HAVE — IIS guard restructured as a positive wrapping conditional.** `if ( ! empty( $is_iis7 ) ) { ...body... }` instead of `if ( empty( $is_iis7 ) ) { return; }`. A bare `return` inside the shared `_imagify_new_upgrade()` exits the whole function and would silently skip any sibling `version_compare` block appended later (harmless today as the last block after 2.2.2, but a real latent defect). The positive wrapper keeps the block self-contained.

3. **SHOULD_HAVE (docs only) — partial remove-then-add failure** documented as a benign, self-healing degraded state (one missing `mimeMap` until the next settings toggle re-fires `add()`), never a fatal 500. Existing `is_wp_error()` guards already cover it; no new code.

**Preserved locked decisions:** dedupe = Option B (`fileExtension`-based); `<preConditions>` collision out of scope (follow-up #1180); DOM-level integration tests accepted (real IIS validation arranged separately); self-heal migration in `inc/admin/upgrader.php`; fileExtension dedupe pinned before the empty-contents early-return.

**Tests:** new `Tests/Integration/classes/WriteFile/` harness (real temp `web.config` + `DOMDocument`) asserting exactly one `<staticContent>` across fresh/foreign/both-formats/idempotent cases; unit tests on `get_raw_new_contents()` (new `@parent`, leaf `mimeMap`, no wrapping `staticContent`, no `name`); self-heal broken-fixture → healed (both `.webp`+`.avif` mimeMaps in one block), partial-failure, and non-IIS/unwritable graceful-skip.

**Test command:** `composer test-unit -- --filter="IIS" && composer test-integration -- --filter="AbstractIISDirConfFile"`; full `composer run-tests` before merge (HIGH risk).

**PR split (L → 2 vertical slices):** PR 1 root-cause fix in the IIS classes + harness (prevents future duplication, self-contained); PR 2 self-heal migration + tests (remediates already-broken sites, builds on PR 1).

**Comment by Honemo** [comment]
## Challenger review of implementation spec (plan_version 1)

**Verdict: NEEDS_REVISION**

Verified against `classes/WriteFile/AbstractIISDirConfFile.php`, `classes/Webp/IIS.php`, `classes/Avif/IIS.php` — the spec's root-cause description and Option B design match the actual code. Two MUST_HAVE gaps found before implementation should start:

1. **No repair path for already-broken production sites.** The fix (retarget `@parent` to `staticContent`, dedupe by `fileExtension`) only prevents *future* duplication the next time `add()`/`remove()` runs (triggered by a settings toggle in `Webp/Display.php` / `Avif/Display.php`). Sites that are *already* down with two `<staticContent>` siblings today will **still be down** after upgrading the plugin, since nothing re-triggers `insert_contents()` on plugin update. Given this issue is literally about live 500 errors happening right now, the spec should add a migration step using the existing `imagify_network_upgrade` / `imagify_first_install` hook mechanism (see `inc/admin/upgrader.php`) to force a rewrite/cleanup of `web.config` on version bump for IIS installs.
2. **Ambiguous code placement for the new fileExtension-based removal in `insert_contents()`.** The abstract marker-removal loop runs *before* the early `return` for empty `$new_contents` (the `remove()` path). The spec's step 3b says to add the new removal "before inserting fragments" without pinning down whether that's before or after this early return — if placed after, `remove()` silently breaks under Option B (since the `name` marker no longer exists on `mimeMap`). This needs an explicit instruction to place it before the `if ( ! $new_contents ) { return ...; }` guard.

Additional (non-blocking) observations:
- Option B's dedupe-by-`fileExtension` will silently delete a **foreign** pre-existing mimeMap for the same extension on `add()`, with no restore on later `remove()` — a legitimate site config could be permanently lost. Worth a doc/edge-case note even if accepted as a known tradeoff.
- No existing integration test scaffolding exists for `WriteFile`/IIS classes (`Tests/Integration/classes/WriteFile/...` doesn't exist yet). Building it from scratch (temp `web.config`, `$is_iis7`/filesystem mocking, real `DOMDocument` + `saveDomDocument()`) is more effort than a typical "M" ticket already carrying 2 concrete-class changes + 1 abstract-engine change; worth confirming the estimate holds once that harness is scoped.

Files referenced: `classes/WriteFile/AbstractIISDirConfFile.php`, `classes/Webp/IIS.php`, `classes/Avif/IIS.php`, `inc/admin/upgrader.php`, `classes/Webp/Display.php`.

— Automated grooming challenger (Claude Sonnet 5)

**Comment by Honemo** [comment]
## Challenger re-review — Plan v2 (round 2)

**Verdict: NEEDS_REVISION**

Both v1 MUST_HAVE findings are correctly resolved:
- The self-heal migration (Part 2, `inc/admin/upgrader.php`) now exists and is version-gated.
- The fileExtension dedupe insertion point is pinned precisely between the marker-removal loop and the empty-contents early return — verified against the actual current source (`AbstractIISDirConfFile.php` lines 36–47 match exactly what the spec describes).

Two new MUST_HAVE gaps found while verifying the spec against the real codebase:

**Finding 1 — MUST_HAVE:** The self-heal migration's re-add step (Part 2, step 7e) assumes `optimization_format` gates which writer (`webp` vs `avif`) should be re-added ("`optimization_format === 'avif'` -> `$avif->add()`, otherwise `$webp->add()`"). This assumption is **false** in the current codebase. `classes/Webp/Display.php::activate()`/`maybe_add_rewrite_rules()` and `classes/Avif/Display.php::activate()`/`maybe_add_rewrite_rules()` are both registered as independent subscribers (`classes/Webp/ServiceProvider.php`, `classes/Avif/ServiceProvider.php`) and both gate purely on `display_nextgen` — neither checks `optimization_format` at all (confirmed: zero matches for `optimization_format` in either Display class). So today, whenever `display_nextgen` is truthy, **both** webp and avif `add()` calls fire together, regardless of `optimization_format`. This is in fact very likely the actual real-world trigger of the reported bug (both classes racing to prepend their own `<staticContent>` simultaneously), and it's exactly what the spec's own Edge Cases table calls out as the normal "Both WebP and AVIF enabled" case. The migration's exclusive-or re-add will under-restore state for any site where `display_nextgen` is on — it will heal the duplicate-`<staticContent>` crash but leave only one of the two expected mimeMaps in place, diverging from steady-state behavior that Part 1's own fix produces for a live settings toggle. Fix: mirror the *actual* activate() gating — if `display_nextgen` is truthy, call `add()` unconditionally on **both** `$webp` and `$avif` (matching current runtime behavior), not an if/else keyed on `optimization_format`.

**Finding 2 — MUST_HAVE:** The IIS guard in the migration block (step 7a) is specified as `global $is_iis7; if ( empty( $is_iis7 ) ) { return; }`, with the explicit instruction "place the guard so it only short-circuits this migration block, not the whole upgrade function." As written, a bare `return;` inside `_imagify_new_upgrade()` exits the **entire function**, not just the new `version_compare` block — this contradicts the spec's own stated requirement in the same sentence. `_imagify_new_upgrade()` is a single flat function with multiple sequential `if ( version_compare(...) )` blocks (verified: current last block is `2.2.2`); today this new block would be appended last so the bug is latent, but it is a real logic defect a developer implementing the spec literally would introduce, silently breaking every future migration block appended after it for non-IIS sites. Fix: wrap the guard as a positive conditional (`if ( ! empty( $is_iis7 ) ) { ...migration body... }`) rather than an early `return`, so it cannot affect sibling code regardless of where the block is inserted relative to future migrations.

**Other requested checks:**
- IIS DOM-merge test plan: reasonable given DOM-level tests can't validate real IIS parsing; the integration test matrix (fresh file, foreign staticContent, foreign mimeMap collision, both formats, idempotent re-add, remove-after-merge, malformed file) covers what's practically verifiable. No further action needed here.
- Remove-then-add atomicity: each `insert_contents()` call is internally atomic (one DOM mutation + one file write), but the *sequence* `webp->remove()` → `avif->remove()` → conditional `add()` is not transactionally atomic across calls. If `webp->remove()` succeeds (file written) but a later call fails (e.g. permissions revoked mid-request), the failure is non-fatal (per step f) and the site is left in a temporary "under-restored" state (missing a mimeMap) until the next successful settings toggle — not a 500, but this partial-failure outcome is not written into the Edge Cases table. Recommend documenting it explicitly (SHOULD_HAVE, not blocking).
- Effort L and the two-slice PR split are reasonable and well-scoped — PR1 (root-cause fix + test harness) is independently mergeable, PR2 (migration) correctly depends on PR1's dedupe behavior. No change needed here.

**Alternative suggestions:**
- For Finding 1, the simplest fix is to drop the `optimization_format` branch entirely and always call both `$webp->add()` and `$avif->add()` when `display_nextgen` is truthy (mirroring the actual unconditional gating in `Webp/Display::activate()` / `Avif/Display::activate()`), then let the Part 1 dedupe/merge logic handle collapsing into one `<staticContent>` as normal.
- For Finding 2, restructure the guard as `if ( ! empty( $is_iis7 ) ) { ... }` wrapping the whole migration body instead of an early `return`, consistent with how other version blocks in `_imagify_new_upgrade()` are self-contained.

**Comment by Honemo** [comment]
## Challenger review — Plan v3 (final round)

**Verdict: APPROVED**

Both round-2 MUST_HAVE fixes were re-verified directly against the current codebase (not taken on the spec's word):

1. **Optimization-format gating (Migration re-add, step 7e).** Confirmed `classes/Webp/Display.php` and `classes/Avif/Display.php` — neither reads `optimization_format` anywhere in the file. Both `activate()` (lines ~123-134 Webp, equivalent Avif) and `maybe_add_rewrite_rules()` gate exclusively on `display_nextgen`, and both fire `$conf->add()` when it is truthy. The spec's v3 text now calls `$webp->add()` AND `$avif->add()` unconditionally when `display_nextgen` is truthy, with no `optimization_format` branch — this matches the real code path. Fixed correctly.

2. **IIS guard structure (step 7a).** Read `inc/admin/upgrader.php::_imagify_new_upgrade()` in full (lines 183-327): it is one shared function with ~10 sequential `version_compare` blocks, the last being 2.2.2. An early `return` inside any new block would indeed have silently skipped nothing today (it's last) but would be a landmine for the next migration author. The v3 spec text now explicitly mandates `if ( ! empty( $is_iis7 ) ) { ...body... }` and explicitly forbids `if ( empty( $is_iis7 ) ) { return; }`, with the rationale spelled out. Fixed correctly.

**Broader final pass (root-cause fix, Part 1):** Traced `AbstractIISDirConfFile::insert_contents()` line-by-line (lines 25-66) — the spec's cited line numbers (36-42 marker-removal, 44-47 early-return, 49-65 fragment insertion) match the file exactly, and the proposed dedupe insertion point (between 42 and 44) is correct for covering both `add()` and `remove()` paths. Also traced `get_node()`/`prepend_node()` (lines 238-282): confirmed the recursion creates a container only when the queried path (`/configuration/system.webServer/staticContent` post-fix) doesn't already resolve, and prepends into an existing node otherwise — this validates the root-cause claim that retargeting `@parent` to the leaf `<mimeMap>` naturally collapses to one shared `<staticContent>`. `TAG_NAME` constants confirmed distinct per class (`Imagify: webp file type` / `Imagify: avif file type` / two separate rewrite-rule markers), so the `<preConditions>` out-of-scope carve-out is correctly isolated from the `<staticContent>` fix. Test paths/commands (`Tests/Integration`, `Tests/Unit`, `composer test-unit`/`test-integration`) match existing repo conventions.

No new MUST_HAVE or SHOULD_HAVE findings. Locked decisions (Option B, #1180 out of scope, user-arranged IIS validation, effort L) not re-litigated. Implementation may proceed.

**Comment by Honemo** [comment]
> [!NOTE]
> Generated by the AI delivery pipeline (orchestrator · Claude Sonnet 5).

**PR:** [#1181](https://github.com/wp-media/imagify-plugin/pull/1181) | **Status:** DRAFT — held pending human review + real IIS validation (by request)

| Stage | Result | Notes |
|---|---|---|
| Grooming | ✅ | effort: L, risk: HIGH (3 revision rounds) |
| Challenger | ✅ Approved (v3) | 2 revision rounds; found real gating bugs, both fixed |
| Implementation | ✅ | branch: fix/509-check-if-rules-already |
| DOD L2 | ✅ PASS | 1 non-blocking warning (Codacy diff-coverage) |
| Lead Review | ✅ PASS | details on PR #1181 |
| CI | ✅ All Pass | |
| QA | ✅ PASS | details on PR #1181 |
| Follow-up tickets | [#1180](https://github.com/wp-media/imagify-plugin/issues/1180) | related `<preConditions>` singleton collision, out of scope here |

**Comment by Honemo** [comment]
~@MathieuLamiot How can we make sure that the file was successfully transmitted and that the CI worked properly?~

**Comment by MathieuLamiot** [comment]
@Honemo I don't understand the question, sorry. Can you elaborate?

**Comment by Honemo** [comment]
@MathieuLamiot Sorry, I posted the comment on the wrong PR.

**Comment by Honemo** [comment]
@tripflex 
Hello, could you please verify that this temporary release fixes your issue? Please note that this is not an official version of the plugin, but a temporary release intended to confirm that the issue has been resolved.

[imagify-iis-fix-rc.zip](https://github.com/user-attachments/files/30129944/imagify-iis-fix-rc.zip)

cc @Miraeld

**Comment by saranshj4** [comment]
Hello @PlanetThoughts,

Would it be possible to try the patched version shared above by developers to confirm the fix for this issue? As this is not official version, it is recommended to try on staging environment
